src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,219 @@
+                                                    item.TargetMethod.MetadataName));
+                                        }
+                                    }
+                                }
+                            }
+                        },
+                        SymbolKind.NamedType);
+                });
+
+            /// <summary>
+            /// Traverse every invocation in the method to find all the dangerous function it invokes.
+            /// </summary>
+            /// <param name="methodSymbol">The Symbol of the method which is going to be analyzed</param>
+            /// <param name="compilation">Compilation context</param>
+            /// <param name="dangerousMethodSymbols">Symbols reprensent dangerous methods</param>
+            /// <param name="visitedMethodSymbols">Methods that have been analyzed</param>
+            IEnumerable<IInvocationOperation> FindDangerousMethodInvocatioinOperation(
+                IMethodSymbol methodSymbol,
+                Compilation compilation,
+                ImmutableArray<IMethodSymbol> dangerousMethodSymbols,
+                ref List<IMethodSymbol> visitedMethodSymbols)
+            {
+                var result = new List<IInvocationOperation>();
+
+                // Every method in the source just needs to be analyzed once.
+                if (visitedMethodSymbols.Find((IMethodSymbol m) => m.Equals(methodSymbol)) == null)
+                {
+                    var blockOperation = methodSymbol.GetTopmostOperationBlock(compilation);
+
+                    if (blockOperation != null)
+                    {
+                        foreach (var operation in blockOperation.Descendants())
+                        {
+                            if (operation.Kind.Equals(OperationKind.Invocation))
+                            {
+                                var invocationOperation = operation as IInvocationOperation;
+                                var invokedMethodSymbol = invocationOperation.TargetMethod;
+
+                                // If it calls a dangerous method, add the invocation operation to the result.
+                                if (dangerousMethodSymbols.Contains(invokedMethodSymbol))
+                                {
+                                    result.Add(invocationOperation);
+                                }
+
+                                // If it calls a method which is reachable, analyze that method recursively.
+                                if (invokedMethodSymbol.IsInSource())
+                                {
+                                    result.AddRange(FindDangerousMethodInvocatioinOperation(invokedMethodSymbol, compilation, dangerousMethodSymbols, ref visitedMethodSymbols));
+                                }
+                            }
+                        }
+                    }
+                }
+
+                visitedMethodSymbols.Add(methodSymbol);
```

Comment:
Should add before visiting.  What happens if you're looking at a method that calls itself?